File: ui/src/types/bots.ts
// This file defines TypeScript types for bot-related data.

export interface BotIdentifier {
  id: string;
  mode: string;
  status: string;
}

export interface Position {
  symbol: string;
  quantity: number;
  price: number;
  side: 'long' | 'short';
}

export interface BotStatus {
  pnl: number;
  positions: Position[];
  balance: number;
  equity: number;
}

export interface Trade {
  timestamp: string;
  symbol: string;
  side: string;
  price: number;
  quantity: number;
  type?: string; // 'open', 'close', or 'unknown'
  pnl?: number; // P&L for closing trades
}

export interface RiskEvaluationDetails {
  value?: number;
  threshold?: number;
  message?: string;
  [key: string]: unknown;
}

export interface RiskEvaluation {
  rule_name: string;
  is_violated: boolean;
  details: RiskEvaluationDetails;
}

export interface BotRisk {
  evaluations: RiskEvaluation[];
  kill_switch_activated: boolean;
}

export interface BotConfig {
  id: string;
  strategy: string;
  symbol: string;
  mode: string;
  initial_balance: number;
  status: string;
  parameters: {
    [key: string]: string | number | boolean;
  };
  risk_settings?: {
    max_position_size?: number;
    max_drawdown?: number;
    [key: string]: unknown;
  };
}

export interface BotConfigUpdate {
  strategy?: string;
  symbol?: string;
  mode?: string;
  initial_balance?: number;
  parameters?: {
    [key: string]: string | number | boolean;
  };
  risk_settings?: {
    max_position_size?: number;
    max_drawdown?: number;
    [key: string]: unknown;
  };
}

File: server/api/schemas/bots.py
# This file defines the Pydantic schemas for bot-related API endpoints.
from pydantic import BaseModel
from typing import List, Dict, Any

class BotIdentifier(BaseModel):
    id: str
    mode: str
    status: str

class BotStatus(BaseModel):
    pnl: float
    positions: List[Dict[str, Any]]
    balance: float
    equity: float

class Trade(BaseModel):
    timestamp: str
    symbol: str
    side: str
    price: float
    quantity: float
    type: str = "unknown"  # "open", "close", or "unknown"
    pnl: float = 0.0  # P&L for closing trades

class RiskEvaluation(BaseModel):
    rule_name: str
    is_violated: bool
    details: Dict[str, Any]

class BotRisk(BaseModel):
    evaluations: List[RiskEvaluation]
    kill_switch_activated: bool

class CreateBotRequest(BaseModel):
    id: str
    strategy: str = "SMA Crossover"
    symbol: str = "BTCUSDT" 
    mode: str = "paper"  # "paper", "simulated", "live"
    initial_balance: float = 10000.0
    parameters: Dict[str, Any] = {
        "fast_period": 20,
        "slow_period": 50,
        "position_size": 0.5
    }
